### PR TITLE
Improve debugging experience

### LINKED ISSUE #97: ENH: Export value function evaluated at states for debugging
#### Is your feature request related to a problem?

Currently, lcm does not throw errors in cases where it should do so. When inspecting the results of a successful run (or even one with errors), it would be extremely helpful to have more information about the value function than just the result given optimal choices. In particular, it would be helpful to be able to look at the value function evaluated at all state-choice combinations.

#### Describe the solution you'd like

Already discussed this shortly with @timmens. He proposed, as a "easy" to implement interim solution, to let lcm export a `.pkl` file that the user could read in to explore unexpected behavior.

Maybe one could combine this with the already existing `debug_mode` argument of `get_lcm_function`. The user could specify a path, where the debug file is dropped. 

#### API breaking implications

None, except potentially changes to the `debug_mode` argument.

#### Describe alternatives you've considered

Alternatives regarding the export itself may be considered. For example, the lcm function may also return the whole array (or only parts of it, specified by the user, e.g. all alternative-specific value functions for some specific choice) as a function output, instead of saving a `.pkl` file to the drive. Then, users do not have to read it in themselves. Given that this is an interim feature, I'm indifferent regarding the two options.

### LINKED ISSUE #142: ENH: Improve logging and debugging
Currently, PyLCM only provides a `debug_mode: bool` flag, which activates status printing.

I suggest the following:

Conceptually, we should differentiate between being ***verbose***, and ***logging***. In the former, PyLCM will try to communicate what it is doing when, with an estimate of how long it will take. In the latter, PyLCM will try to save as much information per period as possible, in order to facilitate ex-post usage of interim results or debugging by the user.

**Verbose:**

1. We create a new boolean flag called `verbose`, which is used as follows
2. We use [tqdm](https://github.com/tqdm/tqdm) to show progress bars:
      ```python
     # in solve_brute.py
      ...
      reversed_periods_range = tqdm(
          reversed(range(n_periods)),
          desc="Solution progress",
          total=n_periods,
          disable=not verbose,
      )
  
      for period in reversed_periods_range:
          ...
       ```

**Logging:**

1. We add a new argument called `logging`, which controls whether data is written to disc or not. This should be:
    - `bool:` Logging is either deactivated (`False`), or activated with a generated log-path (`True`)
    - `str | Path:` Logging is activated and the file will be written to the path (if parents exist, otherwise error)
2. Each period, we write the value function array to disc. Here we need to check:
    - How big can these arrays get for large models
    - What is the best file-format for n-dimensional JAX arrays
    - How do we also store the information on which state corresponds to which axis (potentially use [xarray](https://docs.xarray.dev/en/stable/index.html))

### PR DESCRIPTION
**Pull request by hmgaudecker**
Fixes #97.
Fixes #142 

Replace the boolean `debug_mode` flag with a structured logging/persistence system
that makes it easier to diagnose solve and simulation failures — especially NaN value
functions during parameter estimation.

### Logging rework

`debug_mode: bool` → `log_level: LogLevel` on `solve()`, `simulate()`, and
`solve_and_simulate()`. Four levels:

| Level        | Console output                  | Snapshots |
| ------------ | ------------------------------- | --------- |
| `"off"`      | Nothing                         | No        |
| `"warning"`  | NaN/Inf warnings only           | No        |
| `"progress"` | Per-age timing + regime counts  | No        |
| `"debug"`    | V-function stats per regime-age | Yes       |

Per-period timing in both solve and simulate (elapsed seconds, active regime count).

### Debug snapshots

When `log_level="debug"`, complete snapshots are saved to `log_path`:

```
solve_snapshot_000/
├── arrays.h5          # V_arr_dict in HDF5
├── model.pkl          # cloudpickle'd Model
├── params.pkl         # user params
├── metadata.json      # snapshot type, fields, platform
├── pixi.lock          # environment lock file
└── REPRODUCE.md       # instructions for reproducing
```

Three snapshot types: `SolveSnapshot`, `SimulateSnapshot`,
`SolveAndSimulateSnapshot`. Load with `load_snapshot(path, exclude=(...))` to
skip large fields.

Automatic retention: `log_keep_n_latest=3` (default) deletes oldest snapshots.

### Better error messages

- `InvalidValueFunctionError` now includes regime name and NaN count
  (`"3 of 100 values are NaN in regime 'working_life'"`)
- NaN/Inf in value function arrays logged as warnings before the hard error
- Value function debug stats: min/max/mean per regime per age at debug level

### Validation improvements

- Feasibility check now vmaps over action combos (consistent with solve/simulate),
  fixing shape mismatches with `MappingLeaf` parameters
- Batched vmapped feasibility to cap memory at ~256 MB
- Enhanced infeasibility error: shows a DataFrame of state values with discrete
  codes decoded to labels, plus constraint details and grid bounds

### Other

- `SimulationResult.to_pickle()` / `.from_pickle()` for serialisation
- `save_solution()` / `load_solution()` for value function arrays
- Test file splits for speed in parallel runs (model set up is expensive, so make
  batches by module): 
  - `test_solution_on_toy_model.py` →
    `test_solution_on_toy_model_deterministic.py` +
    `test_solution_on_toy_model_stochastic.py` (speed when 
  - `test_shock_grids.py` → grid tests stay, draw tests split into
    `test_shock_draw.py`
- Test model factories cached with `@functools.cache` for faster test suites

### Documentation

- New user guide page: *Debugging* (`docs/user_guide/debugging.md`) — log levels,
  snapshot structure, NaN diagnosis recipes, value function inspection patterns

### CODE REVIEW

File: src/lcm/persistence.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+3. Load the snapshot:
+
+```python
+from lcm import load_snapshot
+snapshot = load_snapshot("{snap_dir.name}")
+# Re-run: snapshot.model.solve(snapshot.params)
+```
+
+Platform: {platform_str}
+"""
```
Comment (timmens) [inline src/lcm/persistence.py]: Looks a bit weird. Don't you use dedent for that?

File: src/lcm/model.py [hunk 2]
```diff
@@ -242,17 +274,28 @@ def simulate(
...
+                params=params,
+                initial_conditions=initial_conditions,
+                V_arr_dict=V_arr_dict,
+                result=result,
+                log_path=Path(log_path),
+                log_keep_n_latest=log_keep_n_latest,
+            )
+        return result
 
     def solve_and_simulate(
```
Comment (timmens) [inline src/lcm/model.py]: 1. If the log level is on debug; will solve_and_simulate log both the SolutionSnapshot and the SimulationSnapshot, and is that not redundant then?
2. Also here thinking again, we should probably merge solve_and_simulate and simulate.

File: src/lcm/solution/solve_brute.py
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply src/lcm/solution/solve_brute.py]: Done.

File: tests/test_ages.py [hunk 3]
```diff
@@ -163,13 +163,14 @@ def test_model_with_quarterly_steps():
...
     # Solve and simulate
-    result = model.solve_and_simulate(
+    result = model.simulate(
         params=params,
         initial_conditions={
             "wealth": jnp.array([50.0, 100.0, 150.0]),
             "age": jnp.array([18.0, 18.0, 18.0]),
             "regime": jnp.array([RegimeId.working_life] * 3),
         },
+        V_arr_dict=None,
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply tests/test_ages.py]: Not getting rid of the default was an omission on my part, sorry. I agree it is ugly in the tests, but for actual usage there I think it is a big plus of having to pass `None` explicitly, now that we just have `simulate`. I don't want to imagine how many calculations would be wasted because somebody thinks that magically, a pre-computed value function array will be used...

File: src/lcm/solution/solve_brute.py [hunk 4]
```diff
@@ -60,11 +64,31 @@ def solve(
...
+                logger=logger,
+                regime_name=name,
+                age=ages.values[period],
+                V_arr=V_arr,
+            )
+            log_vf_stats(logger=logger, regime_name=name, V_arr=V_arr)
+
+            validate_value_function_array(
+                V_arr=V_arr, age=ages.values[period], regime_name=name
+            )
```
[the remark this answers is not in the document]
Comment (hmgaudecker) [reply src/lcm/solution/solve_brute.py]: Thanks, great point — glad you brought it up. This is worth getting right since the naming touches the whole codebase.

**Status quo — three conventions coexist:**

The codebase uses `X` / `X_arr` consistently for other math objects (`Q_arr`, `U_arr`, `F_arr` for computed arrays; `Q_and_F`, `U_and_F` for callables). But for V, three different styles coexist:

| Style | Examples | Where |
|---|---|---|
| `V_arr` (math) | `V_arr`, `V_arr_dict`, `next_V_arr`, `VArrMapping` | Variables, params, type alias, persistence |
| `value_function` (spelled out) | `validate_value_function_array`, `get_value_function_representation` | error_handling.py, function_representation.py |
| `vf` (abbreviation) | `log_vf_nan`, `log_vf_stats` | logging.py |

**Proposed renames — variables:**

| Current | Proposed | Rationale |
|---|---|---|
| `V_arr` | keep | Matches `Q_arr`, `F_arr` |
| `next_V_arr` | `next_regime_to_V_arr` | Follow `x_to_y` mapper convention |
| `V_arr_dict` | `period_to_regime_to_V_arr` | Same convention, makes nesting explicit |
| `VArrMapping` | `PeriodToRegimeToVArr` | Align type alias with variable name |

**Proposed renames — functions:**

| Current | Proposed |
|---|---|
| `validate_value_function_array()` | `validate_V()` |
| `get_value_function_representation()` | `get_V_representation()` |
| `log_vf_nan()` | `log_nan_in_V()` |
| `log_vf_stats()` | `log_V_stats()` |

Any thoughts?

[the remark this answers is not in the document]
Comment (timmens) [reply src/lcm/solution/solve_brute.py]: I like it! Would be fine with me to go ahead with this.
